docs/what-is-wrapped.md

```diff
@@ -0,0 +1,1007 @@
+    // oops, it doesn't work anymore. We will deal with it later
+    //User(
+    //  id,
+    //  Age(21),
+    //  Email("user@email.com")
+    //)
+
+  def save(user: User): Unit = ()
+}
+
+class UserService(repo: UserRepository, laws: LawsInfoService, emailService: EmailService) {
+  def get(id: UserId): User = repo.get(id)
+  
+  def save(user: User): Unit = {
+    repo.save(user)
+    val message = if (laws.canHaveRussianPassport(user.age)) "hello big buddy, you are saved" else "hello, you are saved"
+    emailService.sendEmail(user.email, message)
+  }
+}
+```
+
+We almost there! `Wrapped` trait now looks exactly the same as in real project.
+There are only two new issues to solve (this is the last two).
+1. Returned error is not flexible enough. What if someone wants to return (not throw!) an exception? What if someone wants to return several errors instead of just one?
+2. We can't just create a new object in repository anymore. Which is actually a problem, because our repository works with a trusted source of data.
+   We should have a way to tell Scala that sometimes it is ok to let us unsafely create a wrapper and _we_ know that it's going to work fine.
+
+Of course, we can just "fix" problem #1 by saying that most of the time we can represent errors as String so people should just call `.toString` on their errors.
+However, we can do better than this and allow people to configure error type using the abstract type.<br>
+Problem #2, however, a bit more interesting. We have made a hard restriction on how we can create a `Wrapper`s, but now we have to lax it and create a workaround to give people a way to create wrapper instances unsafely.
+Let's do it by introducing new method that will allow unsafe creation. However, will do this method so ugly, that it would be impossible to use this method without being noticed.
+Yes, it would be unsafe and an obvious vulnerability, but hopefully it would be a controlled vulnerability that everyone is aware of.
+
+```scala
+trait Wrapped[T] extends Any {
+  def value: T
+  override def toString: String = value.toString
+
+  def canEqual(that: Any): Boolean = this.getClass.isInstance(that)
+  override def equals(that: Any): Boolean = canEqual(that) && this.value.equals(that.asInstanceOf[Wrapped[T]].value)
+  override def hashCode: Int = this.getClass.hashCode + value.hashCode()
+}
+
+object Wrapped {
+  trait Companion {
+    type Type
+    type Wrapper <: Wrapped[Type]
+    type Error
+    type ValidationResult[A] = Either[Error, A]
+ 
+    protected def validate(value: Type): ValidationResult[Type]
+    protected def create(value: Type): Wrapper
+
+    final def from(value: Type): ValidationResult[Wrapper] = validate(value).map(create)
+  
+    final def uglyUnsafeCreation(value: Type): Wrapper = from(value) match {
+      case Right(wrapper) => wrapper
+      case Left(error) => throw new RuntimeException(s"Oh no! You promised me that it's ok to unsafely wrap value but I got an error! $error")
```

Comment:
That one is still vague. How it helps to create unsafely if it still returns only a valid result. And yes, you promised that it's ok to unsafely wrap value but we got an error!

Reply:
"unsafe" here doesn't mean we allow creation with invalid data. Thanks for noticing it, I've explained it in detail in updated document. Please take a look.